### PR TITLE
Skip iptables masquerading for packets destined to remote nodes

### LINKED ISSUE #15403: Unnecessary iptables masquerade from pod to remote host in direct routing mode
It was observered, that when running with iptables masquerade, the following rule will masquerade traffic sent from a pod to a remote Cilium host:
```
[3:252] -A CILIUM_POST_nat -s 10.154.2.0/24 ! -d 10.154.0.0/16 ! -o cilium_+ -m comment --comment "cilium masquerade non-cluster" -j MASQUERADE
```

Not only the masquerade is unnecessary in the direct routing mode, but also it might break the host-fw.

The BPF-based masquerade does not SNAT such packets.

### PR DESCRIPTION
**Pull request by pchaigno**
This pull requests updates the iptables rules installed by Cilium to skip masquerading for traffic destined to cluster nodes. As a summary of commits:
1. Add the `ipset` binary to the Cilium agent image.
2. Update cilium-builder and cilium-runtime.
2. Store node IP addresses in an ipset and use it to skip masquerading.
3. Collect ipsets in the bugtool reports.
4. Workaround for VMs where Cilium runs as a service, until https://github.com/cilium/packer-ci-build/pull/278 is merged.

Fixes: https://github.com/cilium/cilium/issues/15403.

### CODE REVIEW

**Comment by pchaigno** [comment]
test-gke

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-1.16-netnext

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-gke

**Comment by pchaigno** [comment]
test-1.21-4.9

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-1.21-4.9

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

File: Vagrantfile [hunk 1]
```diff
@@ -90,6 +90,10 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
+# Temporary fix until https://github.com/cilium/packer-ci-build/pull/278 is
+# merged. Necessary as Cilium now needs ipset for iptables masquerading.
+sudo apt-get install -y ipset
+
```
Comment (nbusseneau) [inline Vagrantfile:99]: Is there anything blocking https://github.com/cilium/packer-ci-build/pull/278?

**Review by nbusseneau — APPROVED** [review]

File: Vagrantfile (hunk 1 above)
Comment (pchaigno) [reply Vagrantfile:99]: No. I'd just rather have this merged before updating all VM images, or at least be fairly certain this PR will be accepted :slightly_smiling_face:

File: pkg/datapath/iptables/iptables.go [hunk 2]
```diff
@@ -1035,12 +1046,56 @@ func (m *IptablesManager) installForwardChainRulesIpX(prog iptablesInterface, if
...
+	ipset.runProg(progArgs, false)
+}
+
+// RemoveFromBodeIpset removes an IP address from the ipset for cluster nodes.
+func RemoveFromNodeIpset(nodeIP net.IP) {
+	ciliumNodeIpset := ciliumNodeIpsetV4
+	if ip.IsIPv6(nodeIP) {
+		ciliumNodeIpset = ciliumNodeIpsetV6
+	}
+	log.WithField(logfields.IPAddr, nodeIP.String()).Debugf("Removing IP to %s ipset", ciliumNodeIpset)
```
Comment (joamaki) [inline pkg/datapath/iptables/iptables.go]:
```suggestion
	log.WithField(logfields.IPAddr, nodeIP.String()).Debugf("Removing IP %s from %s ipset", nodeIP, ciliumNodeIpset)
```

File: pkg/datapath/iptables/iptables.go [hunk 3]
```diff
@@ -1035,12 +1046,56 @@ func (m *IptablesManager) installForwardChainRulesIpX(prog iptablesInterface, if
...
 
+// AddToNodeIpset adds an IP address to the ipset for cluster nodes. It creates
+// the ipset if it doesn't already exist and doesn't error if either the ipset
+// or the IP already exist.
+func AddToNodeIpset(nodeIP net.IP) {
+	ciliumNodeIpset := ciliumNodeIpsetV4
+	if ip.IsIPv6(nodeIP) {
+		ciliumNodeIpset = ciliumNodeIpsetV6
+	}
+	log.WithField(logfields.IPAddr, nodeIP.String()).Debugf("Adding IP to %s ipset", ciliumNodeIpset)
```
Comment (joamaki) [inline pkg/datapath/iptables/iptables.go]:
```suggestion
	log.WithField(logfields.IPAddr, nodeIP.String()).Debugf("Adding IP %s to %s ipset", nodeIP, ciliumNodeIpset)
```

File: pkg/datapath/iptables/iptables.go [hunk 4]
```diff
@@ -1035,12 +1046,56 @@ func (m *IptablesManager) installForwardChainRulesIpX(prog iptablesInterface, if
...
+	createIpset(ciliumNodeIpset, ip.IsIPv6(nodeIP))
+	progArgs := []string{"add", ciliumNodeIpset, nodeIP.String(), "-exist"}
+	ipset.runProg(progArgs, false)
```
Comment (joamaki) [inline pkg/datapath/iptables/iptables.go]: Shouldn't we handle errors from `runProg` here?

File: pkg/datapath/iptables/iptables.go [hunk 5]
```diff
@@ -1035,12 +1046,56 @@ func (m *IptablesManager) installForwardChainRulesIpX(prog iptablesInterface, if
...
 	snatDstExclusionCIDR, allocRange, hostMasqueradeIP string) error {
+	if option.Config.Tunnel == option.TunnelDisabled {
+		// Exclude traffic to nodes from masquerade.
+		if err := createIpset(prog.getIpset(), prog.getProg() == "ip6tables"); err != nil {
+			return err
+		}
+		progArgs := append(
+			m.waitArgs,
+			"-t", "nat",
+			"-A", ciliumPostNatChain,
```
Comment (joamaki) [inline pkg/datapath/iptables/iptables.go:1090]: Just thinking out loud: should we append or insert here? I would assume here the nat rules are clear when we enter and the masquerading rules are added after this?

File: pkg/datapath/iptables/iptables.go (hunk 2 above)
Comment (pchaigno) [reply pkg/datapath/iptables/iptables.go]: The IP address is already logged with the `WithField(...)` :wink:

File: pkg/datapath/iptables/iptables.go (hunk 5 above)
Comment (pchaigno) [reply pkg/datapath/iptables/iptables.go:1090]: > I would assume here the nat rules are clear when we enter and the masquerading rules are added after this?

Yes, I think that's the assumption. Other rules are appended as well. IIRC, we have code to first clear all old rules (see `removeOldRules`).

File: pkg/datapath/iptables/iptables.go (hunk 2 above)
Comment (joamaki) [reply pkg/datapath/iptables/iptables.go]: Doh. :D

File: pkg/datapath/iptables/iptables.go (hunk 5 above)
Comment (joamaki) [reply pkg/datapath/iptables/iptables.go:1090]: Cool thanks for answering.

File: pkg/datapath/iptables/iptables.go (hunk 2 above)
Comment (joamaki) [reply pkg/datapath/iptables/iptables.go]: Only suggestion then would be `to` -> `from`.

Comment (pchaigno) [reply pkg/datapath/iptables/iptables.go]: Yep, I fixed that :slightly_smiling_face:

**Review by joamaki — APPROVED** [review]
LGTM

**Comment by pchaigno** [comment]
test-me-please

**Review by nebril — APPROVED** [review]

**Comment by borkmann** [comment]
@pchaigno have a chance to rebase?

**Comment by pchaigno** [comment]
The checkpatch failure can be ignored (one empty commit description) and the ConformanceAKS failure is expected, the workflow has been temporarily disabled. The ConformanceAKS workflow was also passing before I rebased.

All team reviews are covered except for cilium/cli. That team's review is only required because of the trivial change to the bugtool so I think it's safe to ignore.

Marking ready to merge.

**Pull request merged**